File: problems/0/1/8/018.hs
-- Copyright (C) 2013 Jorge Aparicio

main :: IO()
main = do
  contents <- readFile "018.in"
  let triangle = map (map read . words) $ lines contents :: [[Int]]
  print . maximum . foldl1 traverse $ triangle

traverse
  :: [Int]  -- previous row
  -> [Int]  -- current row
  -> [Int]  -- traversed row
traverse p = traverse' (0 : p)

traverse' :: [Int] -> [Int] -> [Int]
traverse' [p] [c] = [p+c]
traverse' (p1:p2:ps) (c:cs) = (c + max p1 p2) : traverse' (p2:ps) cs

File: problems/0/1/8/018.rs
// Copyright (C) 2014 Jorge Aparicio

use std::cmp::max;
use std::io::fs::File;
use std::vec_ng::Vec;

fn main() {
    let content = File::open(&Path::new("018.in")).read_to_str().unwrap();

    let costs =
        content.
        lines().
        map(|l| l.words().filter_map(from_str).collect::<Vec<uint>>()).
        fold(vec!(0), |a, b| {
            let n = b.len();
            let mut c = Vec::with_capacity(n);

            for i in range(0, n) {
                if i == 0 {
                    c.push(b.get(i) + *a.get(i));
                } else if i == n - 1 {
                    c.push(b.get(i) + *a.get(i - 1));
                } else {
                    c.push(b.get(i) + *max(a.get(i - 1), a.get(i)));
                }
            }

            c
        });

    println!("{}", *costs.iter().max().unwrap());
}

